# BEYOND TRADITIONAL MAGNETIC RESONANCE PROCESSING WITH ARTIFICIAL INTELLIGENCE * 

Amir Jahangiri ${ }^{\text {à }}$ and Vladislav Orekhov ${ }^{\text {ał }}$<br>andepartment of Chemistry and Molecular Biology, Swedish NMR Centre, University of Gothenburg, Box 465,<br>Gothenburg, 40530, Sweden ${ }^{\dagger}$

May 14, 2024


#### Abstract

Smart signal processing approaches using Artificial Intelligence are gaining momentum in NMR applications. In this study, we demonstrate that AI offers new opportunities beyond tasks addressed by traditional techniques. We developed and trained several artificial neural networks in our new toolbox Magnetic Resonance with Artificial intelligence (MR-Ai) to solve three "impossible" problems: quadrature detection using only Echo (or Anti-Echo) modulation from the traditional Echo/Anti-Echo scheme; accessing uncertainty of signal intensity at each point in a spectrum processed by any given method; and defining a reference-free score for quantitative access of NMR spectrum quality. Our findings highlight the potential of AI techniques to revolutionize NMR processing and analysis.


Keywords NMR $\cdot$ NUS $\cdot \mathrm{AI} \cdot \mathrm{WNN} \cdot$ Quadrature detection[^0]

NMR spectroscopy is a powerful analytical technique widely used to acquire atomic-level information about molecular structure, dynamics, and interactions [1,2]. To derive meaningful insights from the acquired spectra, NMR data processing plays a vital role. Artificial Intelligence (AI), and specifically Deep Learning (DL), presents a compelling alternative to traditional methods in NMR processing [3]. Although early demonstrations of machine learning in NMR date back to the 1970s [4], practical applications have evolved significantly with recent advancements in algorithms and computer hardware. In most cases, DL in NMR data processing focuses on surpassing the existing algorithmic techniques for fast and high-quality solving of traditional tasks such as spectra reconstruction from Non-Uniformly Sampled (NUS) time domain signals [5]-8], virtual homo-decoupling [8--11], spectra denoising [12, 13], and automating peak picking [14, 15]. In this study, we address an intriguing question of whether DL can go beyond the traditional problems and offer new ways of spectra processing and analysis [16], and possibly give us insights for designing new signal processing algorithms [8].

We demonstrate a Magnetic Resonance processing with Artificial intelligence (MR-Ai) solution for the seemingly impossible task of recovering a high-quality spectrum from an incomplete phase-modulated quadrature detection experiment, where only one of the two $\mathrm{P}$ - and $\mathrm{N}$-type parts of phase-modulated quadrature detection experiments is available. Furthermore, we show that MR-Ai is able to perform valuable statistical analyses of spectra reconstructed by other methods and thus provides a new reference-free metric of the spectrum's quality.

Phase-twist lineshape in incomplete quadrature detection as a pattern recognition problem: Traditionally, in multidimensional ( $n \mathrm{D}$ ) NMR experiments, frequency discrimination and obtaining pure, absorptive phase signals rely on quadrature detection. This involves acquiring two data points per time increment and per spectral dimension. For a 2D experiment, where the signal evolves in two-time dimensions $t_{1}$ and $t_{2}$, the amplitude-modulated quadrature detection [17, 18] is implemented by acquiring two separate data sets in the form of cosine and sine modulation:

$$
\begin{aligned}
& \text { Data_set }_{1}\left(\cos _{-} \text {modulated }\right): \cos \left(\Omega_{1} t_{1}\right) \exp \left(i \Omega_{2} t_{2}\right) \\
& \text { Data_set }_{2}(\text { sin_modulated }): \sin \left(\Omega_{1} t_{1}\right) \exp \left(i \Omega_{2} t_{2}\right)
\end{aligned}
$$

where $\Omega_{n}$ is the signal frequency in the $n$th dimension.

In contrast, the phase-modulated data usually obtained from gradient coherence order selection experiments are encoded with frequency as either Echo (P-type data) or Anti-Echo (N-type data) coherence:

$$
\begin{aligned}
& \text { Data_set }_{1}\left(P_{-} \text {type }\right): \exp \left(+i \Omega_{1} t_{1}\right) \exp \left(i \Omega_{2} t_{2}\right) \\
& \text { Data_set }_{2}\left(N_{-} \text {type }\right): \exp \left(-i \Omega_{1} t_{1}\right) \exp \left(i \Omega_{2} t_{2}\right)
\end{aligned}
$$

Individually, each of these datasets produces frequencydiscriminated spectra but exhibits a phase-twist lineshape of the peaks (Fig. 1.a), which is not amenable for normal analysis. Until now, it has been understood that the only way to obtain pure absorptive phase signals in Echo-AntiEcho experiments is by using both $\mathrm{P}$ - and $\mathrm{N}$-type data [19-21]. In this work, we demonstrate that MR-Ai can effectively recognize the twisted lineshapes and convert them into the pure absorption form (Fig. 1.b). To the best of our knowledge, none of the traditional methods demonstrated this capability so far.
![](https://cdn.mathpix.com/cropped/2024_06_04_ba858378e2b4ea3de27dg-02.jpg?height=732&width=804&top_left_y=627&top_left_x=1076)

Figure 1: Illustration of the Echo and normal spectrum. (a) Echo spectrum with a phase-twist lineshape, (b) a normal spectrum with a pure, absorptive phase in the frequency domain, and (c and d) their corresponding Virtual Echo presentations in the time domain respectively. In the figures, $P$ and $N$ represent the P-type and N-type data sets, while $\widetilde{P}$ and $\widetilde{N}$ indicate the time reverse and conjugation of P-type and N-type data sets respectively - procedures for transition between the presentations are indicated by arrows.

In our recent publication, we introduced a DNN architecture called WNN, specifically designed to grasp 1D patterns over the entire NMR spectrum in the frequency domain, such as specific patterns of NUS aliasing artifacts and peak multiples in homo-decoupling [8]. Here, we utilize an updated version of our WNN architecture (see Appendix A for more details) capable of capturing 2D patterns, including the phase twisted peaks associated with the $\mathrm{P}$ - (or $\mathrm{N}$-) type data, as a pattern recognition problem in the frequency domain. Fig. 2 a demonstrates the excellent performance of Echo and Anti-Echo reconstructions of ${ }^{1} \mathrm{H}-{ }^{15} \mathrm{~N}$ correlation spectra by MR-Ai on MALT1 ( $\left.45 \mathrm{kDa}\right)$ protein. Similar results for Ubiquitin (7 kDa), Azurin (14 $\mathrm{kDa}$ ), and Tau (disordered, 441 amino acids) are found in Appendix B Figures B. 1 to B. 4

Echo and Anti-Echo reconstruction as a data completion problem: Figure 11illustrates that the task to rectify the phase twist problem encountered in Echo (or Anti-

![](https://cdn.mathpix.com/cropped/2024_06_04_ba858378e2b4ea3de27dg-03.jpg?height=740&width=808&top_left_y=237&top_left_x=236)

Figure 2: Performance of Echo and Anti-Echo reconstruction by using MR-Ai and CS on real data. (a) $2 \mathrm{D}$ ${ }^{1} \mathrm{H}-{ }^{15} \mathrm{~N}$ - TROSY spectra of MALT1 [22] Echo reconstruction using MR-Ai with predicted uncertainty in pink color - The insets show zooming part and corresponding reference with the actual absolute error between reference and reconstruction in red color. Bar graphs (b) represent $R M S D$ as a traditional reference-based evaluation metric and Boxplots (c) represent normalized uncertainty as the intelligent reference-free evaluation metric for comparison reconstructed spectra using MR-Ai and CS for Malt as described in the Appendix A section for additional details.

Echo) spectra can be viewed as a specific case of NUS reconstruction. Indeed, Fig. 1 c and Fig. 1.d show the time domain equivalents of the twisted (Fig. 1.a) and absorptive (Fig. 1 b) line shapes. The time domain presentation, obtained by the two-dimensional inverse Fourier Transform of the real signal shown in Fig. $1, \mathrm{a}, 1, \mathrm{~b}$ and in the following called the Virtual Echo (VE) [23], clearly shows the roles of the P-type and N-type data. There are four regions defined by the signs of $t_{1}$ and $t_{2}$. P-type data corresponds to the upper-right region where both $t_{1}$ and $t_{2}$ are positive $(P)$, and the time-reversed conjugated $\mathrm{P}$ type data corresponds to the lower-left region where both $t_{1}$ and $t_{2}$ are negative $(\widetilde{P})$. Similarly, N-type data corresponds to the region where $t_{1}$ is negative and $t_{2}$ is positive $(N)$, while the time-reversed and conjugated $\mathrm{N}$-type data is in the region where $t_{1}$ is positive and $t_{2}$ is negative $(\widetilde{N})$. Therefore, rectification of the twisted line shape in the frequency domain is equivalent to completing the missing half of the signal in the VE time domain. The problem is similar to spectral reconstruction from NUS data and can be performed using the Compressed Sensing Iterative Soft Thresholding algorithm (CS-IST) [24, 25], a representative traditional NUS reconstruction technique. The CS algorithm maximizes sparsity of the spectrum, thus filling in the missing data to produce the most compact absorptive form of the signal while suppressing the wider and less sparse dispersive features of the twisted signal. Previously, we demonstrated that for performing data completion in the NUS spectra reconstruction, the MR-Ai utilized pattern recognition [8]. In this work, we point out that the seemingly pattern reconstruction problem of rectifying the Echo (or Anti-Echo) twisted line shapes is akin to the data completion in the time domain and can be addressed by traditional algorithms such as CS.

Both reconstruction methods, MR-Ai and CS-IST, reproduce the spectrum with high quality using either P- or N-type data. Fig. 2b (and Fig. A.3) shows a simple spectra quality metric point-to-point $R M S D$ (and $R_{2}^{s}$ ) between the reconstruction and reference spectra for MALT1 (Tau, Azurin, and Ubiquitin). It was demonstrated that the results from the $R M S D$ (and $R_{2}^{s}$ ) metric correspond well to the results obtained using a more extended and advanced set of the NUScon metrics [8, 26]. With its lower $R M S D$ (and higher $R_{2}^{s}$ ), the MR-Ai displays visibly better results compared to the CS for both Echo and Anti-Echo reconstruction. The even better quality score obtained for the reconstruction from time equivalent $50 \%$ NUS experiment (Fig. 2.b) indicates that well-randomized NUS is a better time-saving strategy than acquiring only $\mathrm{N}$ - or P-type data. Fig. 2 b shows that the quality scores for the Anti-Echo reconstruction are higher than for the Echo regardless of the reconstruction method MR-Ai or CS, which is also reproduced for three other proteins shown in Fig. A.3. At first glance, this is a surprising result since from the theory we expect the quality of the reconstructions to be the same for Echo and Anti-Echo. However, in practice, we should note that these are separate experiments with somewhat different pulse sequences [27], which may lead to imbalances between the two spectra. We reproduced the result in simulations (Fig. A.5 b) where the Echo signal had a somewhat lower amplitude than the Anti-Echo. Then, the better result for the Anti-Echo can be explained by the residual unbalanced contribution of the Anti-Echo part in the traditionally processed reference spectrum. This result underlines the value of the reconstruction approach from the individual Echo or Anti-Echo parts in cases of imbalance between the two or if only one can be practically obtained.

Predicting uncertainty of spectrum intensities with DNN: For any physical measurement, such as the intensity at a point in a reconstructed spectrum, estimating the error is equally important for quantitative analysis and distinguishing true signals from noise and artifacts. This task is particularly challenging when using nonlinear processing methods such as CS and DNN, since the RMSD of the baseline noise can no longer be used as a reliable error estimate. Traditionally, this problem is solved by the brute force approach of repeated measures or postexperiment resampling of the data [26, 28]. DNN offers a much more efficient alternative [29]. It is possible to train a network to predict the quality of the results generated by any method [30] by employing the negative log-likelihood (NLL) as the loss function during the training stage.

$$
\begin{aligned}
N L L\left(y_{i} \mid \mu_{i}, \sigma_{i}\right)= & -\log \left(P D F\left(y_{i} \mid \mu_{i}, \sigma_{i}\right)\right) \\
& \approx \frac{\left(y_{i}-\mu_{i}\right)^{2}}{2 \sigma_{i}^{2}}+\log \left(\sigma_{i}\right)
\end{aligned}
$$

where for point $i$ in the spectrum, $y_{i}$, represents the ground truth value; $\mu_{i}$ is the value produced by the applied method, and $\sigma_{i}$ is the uncertainty represented as the standard deviation of a normal distribution. During training, $\sigma_{i}$ is learned, while $y_{i}$ and $\mu_{i}$ are known.

![](https://cdn.mathpix.com/cropped/2024_06_04_ba858378e2b4ea3de27dg-04.jpg?height=363&width=807&top_left_y=740&top_left_x=239)

Figure 3: Illustration of training and predicting MR-Ai for estimation of the uncertainty of the reconstruction generated by any method

The modified MR-Ai architecture employing the NLL as the loss function is illustrated in Fig. 3. In this work, we trained three MR-Ai to predict intensity uncertainties in the Echo (or Anti-Echo) spectra reconstructed by MR-Ai and CS, as well as NUS spectra reconstructed by CS. In Fig. 2 a (and Figures B.5 to B.8), the pink color represents the estimated uncertainty with a $95 \%$ confidence interval (CI) overlaid on the reconstructed spectrum, while the true error is superimposed on the reference spectrum in red for visual comparison.

Reference-free spectrum quality score, pSQ: The predicted uncertainty can be used as a reference-free score of the spectrum quality, akin to the predicted protein structure accuracy pTM-score calculated by AlphaFold [31]. In Fig. 2.c (and Fig. A.4, box plots display the estimated normalized sigma for all points with the mean (green triangles) and median (orange bars). The box plots correlate well with the reference-based scores shown in Fig. 2 b. Moreover, the box plot sores for the Echo and Anti-Echo reconstructions are nearly the same, as they should be, whereas the reference-based scores depicted in Fig. 2.b wrongly prefer the Anti-Echo because of the biased reference as described above.

In this work, we introduce the MR-Ai toolbox, which offers intelligent NMR processing capabilities beyond traditional techniques. The reconstruction of spectra using the incomplete Echo/Anti-Echo quadrature detection pair can be advantageous for saving experimental time and in cases where obtaining a complete and intensity-balanced signal for traditional quadrature is problematic. The predicted uncertainties of spectral intensities and quantitative reference-free spectrum quality metric will aid in the development of new spectrum processing algorithms and may become a crucial component in methods utilizing targeted acquisitions approaches [32, 33]. Our results demonstrate the potential of AI to expand the scope of traditional NMR signal processing and analysis.

## Acknowledgements

The work was supported by the Swedish Research Council grants to 2019-03561 and 2023-03485 to V.O. This study used NMRbox: National Center for Biomolecular NMR Data Processing and Analysis, a Biomedical Technology Research Resource (BTRR), which is supported by NIH grant P41GM111135 (NIGMS).

## References

[1] J. Cavanagh, W. J. Fairbrother, A. G. Palmer III, and N. J. Skelton, Protein NMR spectroscopy: principles and practice. Academic press, 1996.

[2] T. D. Claridge, High-resolution NMR techniques in organic chemistry, vol. 27. Elsevier, 2016.

[3] D. Chen, Z. Wang, D. Guo, V. Orekhov, and X. Qu, "Review and prospect: Deep learning in nuclear magnetic resonance spectroscopy," Chemistry - A European Journal, vol. 26, no. 46, pp. 10391-10401, 2020.

[4] C. Reilly and B. Kowalski, "Nuclear magnetic resonance spectral interpretation by pattern recognition," The Journal of Physical Chemistry, vol. 75, no. 10, pp. 1402-1411, 1971.

[5] X. Qu, Y. Huang, H. Lu, T. Qiu, D. Guo, T. Agback, V. Orekhov, and Z. Chen, "Accelerated nuclear magnetic resonance spectroscopy with deep learning," Angewandte Chemie, vol. 132, no. 26, pp. 1038310386, 2020.

[6] D. Hansen, "Using deep neural networks to reconstruct non-uniformly sampled $\mathrm{nmr}$ spectra," Journal of Biomolecular NMR, vol. 73, 112019.

[7] G. Karunanithy and D. Hansen, "Fid-net: A versatile deep neural network architecture for $\mathrm{nmr}$ spectral reconstruction and virtual decoupling," Journal of Biomolecular NMR, vol. 75, 052021.

[8] A. Jahangiri, X. Han, D. Lesovoy, T. Agback, P. Agback, A. Achour, and V. Orekhov, "Nmr spectrum reconstruction as a pattern recognition problem," Journal of Magnetic Resonance, vol. 346, p. 107342, 2023.

[9] G. Karunanithy, H. W. Mackenzie, and D. F. Hansen, "Virtual homonuclear decoupling in direct detection nuclear magnetic resonance experiments using deep neural networks," Journal of the American Chemical Society, vol. 143, no. 41, pp. 16935-16942, 2021.

Magnetic Resonance processing with Artificial intelligence (MR-Ai)

[10] K. Kazimierczuk, P. Kasprzak, P. S. Georgoulia, I. Matečko-Burmann, B. M. Burmann, L. Isaksson, E. Gustavsson, S. Westenhoff, and V. Y. Orekhov, "Resolution enhancement in nmr spectra by deconvolution with compressed sensing reconstruction," Chemical Communications, vol. 56, no. 93, pp. 14585-14588, 2020.

[11] T. Qiu, A. Jahangiri, X. Han, D. Lesovoy, T. Agback, P. Agback, A. Achour, X. Qu, and V. Orekhov, "Resolution enhancement of $\mathrm{nmr}$ by decoupling with the low-rank hankel model," Chemical Communications, vol. 59, no. 36, pp. 5475-5478, 2023.

[12] H. Lee and H. Kim, "Intact metabolite spectrum mining by deep learning in proton magnetic resonance spectroscopy of the brain," Magnetic Resonance in Medicine, 032019.

[13] D. Chen, W. Hu, H. Liu, Y. Zhou, T. Qiu, Y. Huang, Z. Wang, M. Lin, L. Lin, Z. Wu, et al., "Magnetic resonance spectroscopy deep learning denoising using few in vivo data," IEEE Transactions on Computational Imaging, 2023.

[14] P. Klukowski, M. Augoff, M. Zieba, M. Drwal, A. Gonczarek, and M. Walczak, "Nmrnet: A deep learning approach to automated peak picking of protein nmr spectra," Bioinformatics (Oxford, England), vol. 34, 032018.

[15] D.-W. Li, A. L. Hansen, L. Bruschweiler-Li, C. Yuan, and R. Brüschweiler, "Fundamental and practical aspects of machine learning for the peak picking of biomolecular nmr spectra," Journal of Biomolecular NMR, pp. 1-9, 2022.

[16] V. K. Shukla, G. T. Heller, and D. F. Hansen, "Biomolecular nmr spectroscopy in the era of artificial intelligence," Structure, 2023.

[17] D. States, R. Haberkorn, and D. Ruben, "A twodimensional nuclear overhauser experiment with pure absorption phase in four quadrants," Journal of Magnetic Resonance (1969), vol. 48, no. 2, pp. 286-292, 1982.

[18] D. Marion, M. Ikura, R. Tschudin, and A. Bax, "Rapid recording of $2 \mathrm{~d} \mathrm{nmr}$ spectra without phase cycling. application to the study of hydrogen exchange in proteins," Journal of Magnetic Resonance (1969), vol. 85, no. 2, pp. 393-399, 1989.

[19] A. L. Davis, J. Keeler, E. D. Laue, and D. Moskau, "Experiments for recording pure-absorption heteronuclear correlation spectra using pulsed field gradients," Journal of Magnetic Resonance (1969), vol. 98, no. 1, pp. 207-216, 1992.

[20] G. Kontaxis, J. Stonehouse, E. Laue, and J. Keeler, "The sensitivity of experiments which use gradient pulses for coherence-pathway selection," Journal of Magnetic Resonance, Series A, vol. 111, no. 1, pp. 7076, 1994 .
[21] M. Bostock, D. Holland, and D. Nietlispach, "Improving resolution in multidimensional $\mathrm{nmr}$ using random quadrature detection with compressed sensing reconstruction," Journal of Biomolecular NMR, vol. 68, pp. 67-77, 2017.

[22] X. Han, M. Levkovets, D. Lesovoy, R. Sun, J. Wallerstein, T. Sandalova, T. Agback, A. Achour, P. Agback, and V. Y. Orekhov, “Assignment of ivl-methyl side chain of the ligand-free monomeric human malt1 paracaspase-ig13 domain in solution," Biomolecular NMR Assignments, vol. 16, no. 2, pp. 363-371, 2022.

[23] M. Mayzel, K. Kazimierczuk, and V. Y. Orekhov, "The causality principle in the reconstruction of sparse nmr spectra," Chemical Communications, vol. 50, no. 64, pp. 8947-8950, 2014.

[24] K. Kazimierczuk and V. Y. Orekhov, "Accelerated nmr spectroscopy by using compressed sensing," Angewandte Chemie International Edition, vol. 50, no. 24, pp. 5556-5559, 2011.

[25] D. J. Holland, M. J. Bostock, L. F. Gladden, and D. Nietlispach, "Fast multidimensional $\mathrm{nmr}$ spectroscopy using compressed sensing," Angewandte Chemie International Edition, vol. 50, no. 29, pp. 6548-6551, 2011.

[26] Y. Pustovalova, F. Delaglio, D. L. Craft, H. Arthanari, A. Bax, M. Billeter, M. J. Bostock, H. Dashti, D. F. Hansen, S. G. Hyberts, et al., "Nuscon: a communitydriven platform for quantitative evaluation of nonuniform sampling in $\mathrm{nmr}$," Magnetic Resonance, vol. 2, no. 2, pp. 843-861, 2021.

[27] T. J. Horne and G. A. Morris, "P-type gradientenhanced cosy experiments show lower $t 1$ noise than n-type," Magnetic resonance in chemistry, vol. 35, no. 10, pp. 680-686, 1997.

[28] M. Mayzel, A. Ahlner, P. Lundström, and V. Y. Orekhov, "Measurement of protein backbone 13 co and $15 \mathrm{n}$ relaxation dispersion at high resolution," Journal of Biomolecular NMR, vol. 69, pp. 1-12, 2017.

[29] M. Abdar, F. Pourpanah, S. Hussain, D. Rezazadegan, L. Liu, M. Ghavamzadeh, P. Fieguth, X. Cao, A. Khosravi, U. R. Acharya, V. Makarenkov, and S. Nahavandi, "A review of uncertainty quantification in deep learning: Techniques, applications and challenges," Information Fusion, vol. 76, pp. 243297, 2021.

[30] G. Scalia, C. A. Grambow, B. Pernici, Y.-P. Li, and W. H. Green, "Evaluating scalable uncertainty estimation methods for deep learning-based molecular property prediction," Journal of chemical information and modeling, vol. 60, no. 6, pp. 2697-2717, 2020 .

[31] J. Jumper, R. Evans, A. Pritzel, T. Green, M. Figurnov, O. Ronneberger, K. Tunyasuvunakool,

R. Bates, A. Žídek, A. Potapenko, et al., "Highly accurate protein structure prediction with alphafold," Nature, vol. 596, no. 7873, pp. 583-589, 2021.

[32] V. A. Jaravine and V. Y. Orekhov, "Targeted acquisition for real-time $\mathrm{nmr}$ spectroscopy," Journal of the American Chemical Society, vol. 128, no. 41, pp. 13421-13426, 2006.

[33] L. Isaksson, M. Mayzel, M. Saline, A. Pedersen, J. Rosenlöw, B. Brutscher, B. G. Karlsson, and V. Y. Orekhov, "Highly efficient nmr assignment of intrinsically disordered proteins: application to b-and $t$ cell receptor domains," PLos one, vol. 8, no. 5, p. e62947, 2013.

[34] A. van den Oord, S. Dieleman, H. Zen, K. Simonyan, O. Vinyals, A. Graves, N. Kalchbrenner, A. Senior, and K. Kavukcuoglu, "Wavenet: A generative model for raw audio," 2016.

[35] A. F. Agarap, "Deep learning using rectified linear units (relu)," arXiv preprint arXiv:1803.08375, 2018.

[36] M. Abadi, A. Agarwal, P. Barham, E. Brevdo, Z. Chen, C. Citro, G. S. Corrado, A. Davis, J. Dean, M. Devin, et al., "Tensorflow: Large-scale machine learning on heterogeneous distributed systems," arXiv preprint arXiv:1603.04467, 2016.

[37] D. P. Kingma and J. Ba, "Adam: A method for stochastic optimization," arXiv preprint arXiv:1412.6980, 2014.

[38] M. W. Maciejewski, A. D. Schuyler, M. R. Gryk, I. I. Moraru, P. R. Romero, E. L. Ulrich, H. R. Eghbalnia, M. Livny, F. Delaglio, and J. C. Hoch, "Nmrbox: a resource for biomolecular nmr computation," Biophysical journal, vol. 112, no. 8, pp. 1529-1534, 2017.

[39] F. Delaglio, S. Grzesiek, G. W. Vuister, G. Zhu, J. Pfeifer, and A. Bax, "Nmrpipe: a multidimensional spectral processing system based on unix pipes," Journal of biomolecular NMR, vol. 6, pp. 277-293, 1995.

[40] J. J. Helmus and C. P. Jaroniec, "Nmrglue: an open source python package for the analysis of multidimensional nmr data," Journal of biomolecular NMR, vol. 55, pp. 355-367, 2013.

[41] S. G. Hyberts, K. Takeuchi, and G. Wagner, "Poissongap sampling and forward maximum entropy reconstruction for enhancing the resolution and sensitivity of protein nmr data," Journal of the American Chemical Society, vol. 132, no. 7, pp. 2145-2147, 2010.

[42] P. S. Brzovic, A. Lissounov, D. E. Christensen, D. W. Hoyt, and R. E. Klevit, "A ubch5/ubiquitin noncovalent complex is required for processive brca1directed ubiquitination," Molecular cell, vol. 21, no. 6, pp. 873-880, 2006.
[43] D. M. Korzhnev, B. G. Karlsson, V. Y. Orekhov, and M. Billeter, "Nmr detection of multiple transitions to low-populated states in azurin," Protein Science, vol. 12, no. 1, pp. 56-65, 2003.

[44] S. Unnerståle, M. Nowakowski, V. Baraznenok, G. Stenberg, J. Lindberg, M. Mayzel, V. Orekhov, and T. Agback, "Backbone assignment of the malt1 paracaspase by solution nmr," Plos one, vol. 11, no. 1, p. e0146496, 2016.

[45] D. M. Lesovoy, P. S. Georgoulia, T. Diercks, I. Matečko-Burmann, B. M. Burmann, E. V. Bocharov, W. Bermel, and V. Y. Orekhov, "Unambiguous tracking of protein phosphorylation by fast high-resolution fosy nmr," Angewandte Chemie International Edition, vol. 60, no. 44, pp. 23540-23544, 2021.
